src/celeritas/em/interactor/MuBetheBlochInteractor.hh

```diff
@@ -0,0 +1,257 @@
+    , max_secondary_energy_(this->calc_max_secondary_energy())
+    , use_rad_corr_(inc_energy_ > value_as<Energy>(inc_energy_limit())
+                    && max_secondary_energy_
+                           > value_as<Energy>(secondary_energy_limit()))
+    , envelope_dist_(this->calc_envelope_distribution())
+{
+    CELER_EXPECT(particle.particle_id() == shared_.ids.mu_minus
+                 || particle.particle_id() == shared_.ids.mu_plus);
+    CELER_EXPECT(inc_energy_ > electron_cutoff_);
+    CELER_EXPECT(inc_energy_
+                 >= value_as<Energy>(detail::mu_ionization_limit()));
+}
+
+//---------------------------------------------------------------------------//
+/*!
+ * Simulate discrete muon ionization.
+ */
+template<class Engine>
+CELER_FUNCTION Interaction MuBetheBlochInteractor::operator()(Engine& rng)
+{
+    if (electron_cutoff_ > max_secondary_energy_)
+    {
+        // No interaction if the maximum secondary energy is below the cutoff
+        return Interaction::from_unchanged();
+    }
+
+    // Allocate secondary electron
+    Secondary* secondary = allocate_(1);
+    if (secondary == nullptr)
+    {
+        // Failed to allocate space for a secondary
+        return Interaction::from_failure();
+    }
+
+    real_type secondary_energy;
+    real_type target_dist;
+    do
+    {
+        real_type u = generate_canonical(rng);
+        secondary_energy
+            = electron_cutoff_ * max_secondary_energy_
+              / (electron_cutoff_ * (1 - u) + max_secondary_energy_ * u);
+        target_dist = 1 - beta_sq_ * secondary_energy / max_secondary_energy_
+                      + ipow<2>(secondary_energy)
+                            / (2 * ipow<2>(total_energy_));
+
+        if (use_rad_corr_
+            && secondary_energy > value_as<Energy>(secondary_energy_limit()))
+        {
+            real_type a1 = std::log(1 + 2 * secondary_energy / electron_mass_);
+            real_type a3 = std::log(4 * total_energy_
+                                    * (total_energy_ - secondary_energy)
+                                    / ipow<2>(mass_));
+            target_dist *= (1 + alpha_over_twopi() * a1 * (a3 - a1));
+        }
+    } while (!BernoulliDistribution(target_dist / envelope_dist_)(rng));
+
+    // TODO: If the \c UseAngularGeneratorFlag is set (false by default), use
+    // the angular generator interface to sample the secondary direction (see
+    // \c G4EmParameters::UseAngularGeneratorForIonisation())
+    real_type secondary_momentum = std::sqrt(
+        secondary_energy * (secondary_energy + 2 * electron_mass_));
+    real_type total_momentum = total_energy_ * std::sqrt(beta_sq_);
+    real_type costheta = secondary_energy * (total_energy_ + electron_mass_)
+                         / (secondary_momentum * total_momentum);
+    CELER_ASSERT(costheta <= 1);
+
+    // Sample and save outgoing secondary data
+    UniformRealDistribution<real_type> sample_phi(0, 2 * constants::pi);
```

Comment:
Even better: since we repeat this line and the next in so many places, we really need a helper function for it. Another helper function for some of the energy conservation calculations would be good too...

Reply:
Actually, what did you have in mind for this helper function? Since it's just two lines of code I'm not sure what would help...